File: Pattern/Monotonic Stack/503_Next_Greater_Element_II/Solution.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        ArrayDeque<Integer> stack = new ArrayDeque();
        Arrays.fill(res,-1);
        for(int i = 0 ; i <n*2 ;i++){
            int num = nums[i%n];
            while(!stack.isEmpty() && nums[stack.peek()] < num){
                int idx = stack.pop();
                res[idx] = num;
            }
            if(i < n){
                stack.push(i);
            }
        }
        return res;
    }
}

File: pattern/Monotonic Stack/503_Next_Greater_Element_II/Solution.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        stack = []
        n = len(nums)
        res = [-1]*n
        for i in range(0,n*2):
            num = nums[i%n]
            while len(stack) > 0 and nums[stack[-1]] < nums[i%n]:
                idx = stack.pop()
                res[idx] = nums[i%n]
            if i < n :
                stack.append(i)
        return res
